imap_processing/hi/l2/hi_l2.py

```diff
@@ -0,0 +1,192 @@
+        ENA signal rates computed from the binned PSET data.
+    """
+    signal_rate_vars = {}
+    # Allow divide by zero to fill set pixels with zero exposure time to NaN
+    with np.errstate(divide="ignore"):
+        # Calculate the ENA Signal Rate and Signal Rate Uncertainties
+        signal_rate_vars["ena_signal_rates"] = (
+            map_ds["counts"] / map_ds["exposure_factor"] - map_ds["bg_rates"]
+        )
+        signal_rate_vars["ena_signal_rate_stat_unc"] = np.sqrt(
+            map_ds["counts"] / map_ds["exposure_factor"]
+        )
+    # Statistical fluctuations may result in a negative ENA signal rate after
+    # background subtraction. A negative signal rate is nonphysical. See IMAP Hi
+    # Algorithm Document section 3.1.1
+    signal_rate_vars["ena_signal_rates"].values[
+        signal_rate_vars["ena_signal_rates"].values < 0
+    ] = 0
+    # Minimum count uncertainty is 1 for any pixel that has non-zero exposure time
+    # See IMAP Hi Algorithm Document section 3.1.1
+    signal_rate_vars["ena_signal_rate_stat_unc"].values[
+        (signal_rate_vars["ena_signal_rate_stat_unc"].values == 0)
+    ] = 1
+    return signal_rate_vars
+
+
+def calculate_ena_intensity(
+    map_ds: xr.Dataset,
+    geometric_factors_path: str | Path,
+    esa_energies_path: str | Path,
+) -> dict[str, xr.DataArray]:
+    """
+    Calculate the ena intensities.
+
+    Parameters
+    ----------
+    map_ds : xr.Dataset
+        Map dataset that has ena_signal_rate fields calculated.
+    geometric_factors_path : str or Path
+        Where to get the geometric factors from.
+    esa_energies_path : str or Path
+        Where to get the energies from.
+
+    Returns
+    -------
+    intensity_vars : dict[str, xarray.DataArray]
+        ENA Intensity with statistical and systematic uncertainties.
+    """
+    # TODO: Implement geometric factor lookup
+    if geometric_factors_path is not None:
+        raise NotImplementedError
+    geometric_factor = xr.DataArray(
+        np.ones((map_ds["esa_energy_step"].size, map_ds["calibration_prod"].size)),
+        coords=[map_ds["esa_energy_step"], map_ds["calibration_prod"]],
+    )
+    # TODO: Implement esa energies lookup
+    if esa_energies_path is not None:
+        raise NotImplementedError
+    esa_energy = xr.ones_like(map_ds["esa_energy_step"])
+
+    # Convert ENA Signal Rate to Flux
+    intensity_vars = {
+        "ena_intensity": map_ds["ena_signal_rates"] / geometric_factor / esa_energy,
+        "ena_intensity_stat_unc": map_ds["ena_signal_rate_stat_unc"]
+        / geometric_factor
+        / esa_energy,
+        "ena_intensity_sys_err": map_ds["bg_rates_unc"] / geometric_factor / esa_energy,
+    }
+
+    # TODO: Correctly implement combining of calibration products. For now, just sum
+    intensity_vars["ena_intensity"] = intensity_vars["ena_intensity"].sum(
+        dim="calibration_prod"
+    )
+    intensity_vars["ena_intensity_stat_unc"] = np.sqrt(
```

Comment:
Can you clarify in a comment somewhere what this "calibration_product" axis represents?